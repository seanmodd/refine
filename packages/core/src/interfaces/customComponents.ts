import React from "react";

export type TitleProps = {
    collapsed: boolean;
};

export type LayoutProps = {
    Sider: React.FC;
    Header: React.FC;
    Title: React.FC<TitleProps>;
    Footer: React.FC;
    OffLayoutArea: React.FC;
    dashboard?: boolean;
};
