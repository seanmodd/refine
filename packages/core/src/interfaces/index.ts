// reducers
export * from "./reducers/auth/user";
export * from "./reducers/resource/resource";

// redux state
export * from "./istate";

// contexts
export * from "../contexts/data/IDataContext";
export * from "../contexts/auth/IAuthContext";
export * from "../contexts/refine/IRefineContext";
export * from "../contexts/translation/ITranslationContext";
export * from "../contexts/components/IComponentsContext";
export * from "../contexts/notification/INotificationContext";
export * from "../contexts/resource/IResourceContext";
export * from "../contexts/unsavedWarn/IUnsavedWarnContext";

// buttons
export { CreateButtonProps } from "../components/buttons/create/index";

export * from "../components/pages/login";

// field
export * from "./field";

// notification
export * from "./notification";

// match
export * from "./match";

// mutationMode
export * from "./mutationMode";

// mutationMode
export * from "./HttpError";

// custom components
export * from "./customComponents";

// resourceRouterParams
export * from "./resourceRouterParams";

// resourceErrorRouterParams
export * from "./resourceErrorRouterParams";

// redirection
export * from "./redirectionTypes";

// upload
export * from "./upload";

// mapData
export * from "./mapDataFn";

// successErrorNotification
export * from "./successErrorNotification";

//metaData
export * from "./metaData";

// sunflower types
export {
    useModalFormFromSFReturnType,
    useStepsFormFromSFReturnType,
} from "../../types/sunflower";

export type BaseRecord = {
    id?: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
};
export interface Option {
    label: string;
    value: string;
}
