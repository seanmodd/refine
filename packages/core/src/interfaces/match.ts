export type MatchRoute = {
    params: {
        resourceName: string;
        action: string;
        id: string;
    };
};
