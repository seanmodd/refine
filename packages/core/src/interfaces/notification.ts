export interface INotification {
    id: string;
    resource: string;
    cancelMutation: () => void;
    doMutation: () => void;
    seconds: number;
    isRunning: boolean;
    isSilent: boolean;
}
