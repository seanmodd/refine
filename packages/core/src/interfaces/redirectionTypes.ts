export type RedirectionTypes = "show" | "list" | "edit" | false;
