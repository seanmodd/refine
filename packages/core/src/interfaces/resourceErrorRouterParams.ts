export type ResourceErrorRouterParams = {
    resource: string;
    action: "show" | "edit" | "create" | undefined;
};
