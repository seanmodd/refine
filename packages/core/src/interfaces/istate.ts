import { IAuthUserReducer } from "../interfaces";
import { IResourceReducer } from "./reducers/resource/resource";

export interface IState {
    auth: {
        user: IAuthUserReducer;
    };
    resources: {
        [name: string]: IResourceReducer;
    };
}
