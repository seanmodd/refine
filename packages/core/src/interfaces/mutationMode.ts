import { BaseRecord, GetListResponse, GetOneResponse } from "../interfaces";
import { QueryKey } from "react-query";

export type MutationMode = "pessimistic" | "optimistic" | "undoable";

export type QueryResponse<T = BaseRecord> =
    | GetListResponse<T>
    | GetOneResponse<T>;

export type Context = {
    previousQueries: ContextQuery[];
};

export type ContextQuery<T = BaseRecord> = {
    query: QueryResponse<T>;
    queryKey: QueryKey;
};
