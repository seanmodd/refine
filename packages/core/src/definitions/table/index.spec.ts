import {
    stringifyTableParams,
    parseTableParams,
    getDefaultSortOrder,
    getDefaultFilter,
    mapAntdSorterToCrudSorting,
    mapAntdFilterToCrudFilter,
    unionFilters,
    compareFilters,
} from "./";
import { TablePaginationConfig } from "@components/antd";
import { CrudSorting, CrudFilters } from "../../interfaces";

describe("definitions/table", () => {
    it("stringify table params correctly", async () => {
        const pagination: TablePaginationConfig = {
            current: 1,
            pageSize: 10,
        };

        const sorter: CrudSorting = [
            {
                field: "id",
                order: "desc",
            },
            {
                field: "title",
                order: "desc",
            },
        ];

        const filters: CrudFilters = [
            {
                field: "categoryId",
                operator: "in",
                value: [1, 2],
            },
        ];

        const url = stringifyTableParams({ pagination, sorter, filters });
        expect(url).toMatchSnapshot();
    });

    it("stringify table single sort params correctly", async () => {
        const pagination: TablePaginationConfig = {
            current: 1,
            pageSize: 10,
        };

        const sorter: CrudSorting = [{ field: "id", order: "desc" }];
        const filters: CrudFilters = [
            {
                field: "categoryId",
                operator: "in",
                value: [1, 2],
            },
        ];

        const url = stringifyTableParams({ pagination, sorter, filters });
        expect(url).toMatchSnapshot();
    });

    it("parse table params with single sorter correctly", async () => {
        const url =
            "?current=1&pageSize=10&categoryId__in[]=1&categoryId__in[]=2&sort[]=id&order[]=desc";

        const { parsedCurrent, parsedPageSize, parsedSorter, parsedFilters } =
            parseTableParams(url);

        expect(parsedCurrent).toBe(1);
        expect(parsedPageSize).toBe(10);
        expect(parsedSorter).toStrictEqual([{ field: "id", order: "desc" }]);
        expect(parsedFilters).toStrictEqual([
            { field: "categoryId", operator: "in", value: ["1", "2"] },
        ]);
    });

    it("getDefaultSortOrder", () => {
        const sorter: CrudSorting = [
            {
                field: "title",
                order: "asc",
            },
            {
                field: "view",
                order: "desc",
            },
        ];

        expect(getDefaultSortOrder("title", sorter)).toEqual("ascend");
    });

    it("getDefaultFilter", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "contains",
                value: "test",
            },
        ];
        expect(getDefaultFilter("title", filters, "contains")).toEqual("test");
    });

    it("getDefaultFilter empty array", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "contains",
                value: undefined,
            },
        ];
        expect(getDefaultFilter("title", filters, "contains")).toEqual([]);
    });

    it("getDefaultFilter default operator", () => {
        const filters: CrudFilters = [
            {
                field: "title",
                operator: "eq",
                value: "test",
            },
        ];
        expect(getDefaultFilter("title", filters)).toEqual("test");
    });

    it("mapAntdSorterToCrudSorting", () => {
        expect(
            mapAntdSorterToCrudSorting({
                field: "title",
                order: "descend",
            }),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "title",
                "order": "desc",
              },
            ]
        `);
    });

    it("mapAntdSorterToCrudSorting with sorting priority", () => {
        expect(
            mapAntdSorterToCrudSorting([
                {
                    field: "id",
                    order: "descend",
                    column: {
                        sorter: {
                            multiple: 2,
                        },
                    },
                },
                {
                    field: "title",
                    order: "descend",
                    column: {
                        sorter: {
                            multiple: 1,
                        },
                    },
                },
            ]),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "title",
                "order": "desc",
              },
              Object {
                "field": "id",
                "order": "desc",
              },
            ]
        `);
    });

    it("mapAntdSorterToCrudSorting for array and columnKey", () => {
        expect(
            mapAntdSorterToCrudSorting([
                {
                    columnKey: "title",
                    field: "title",
                    order: "descend",
                },
            ]),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "title",
                "order": "desc",
              },
            ]
        `);
    });

    it("mapAntdFilterToCrudFilter", () => {
        expect(
            mapAntdFilterToCrudFilter(
                {
                    foo: ["bar", "baz"],
                },
                [],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "in",
                "value": Array [
                  "bar",
                  "baz",
                ],
              },
            ]
        `);
    });

    it("mapAntdFilterToCrudFilter with non array", () => {
        expect(
            mapAntdFilterToCrudFilter(
                {
                    foo: "bar",
                },
                [],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "eq",
                "value": "bar",
              },
            ]
        `);
    });

    it("mapAntdFilterToCrudFilter with value 0", () => {
        expect(
            mapAntdFilterToCrudFilter(
                {
                    foo: [0],
                },
                [],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "in",
                "value": Array [
                  0,
                ],
              },
            ]
        `);
    });

    it("mapAntdFilterToCrudFilter with in operator and null value", () => {
        expect(
            mapAntdFilterToCrudFilter(
                {
                    foo: null,
                },
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: ["1"],
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "in",
                "value": null,
              },
            ]
        `);
    });

    it("mapAntdFilterToCrudFilter with eq operator and null value", () => {
        expect(
            mapAntdFilterToCrudFilter(
                {
                    foo: null,
                },
                [
                    {
                        field: "foo",
                        operator: "eq",
                        value: "1",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "foo",
                "operator": "eq",
                "value": null,
              },
            ]
        `);
    });

    it("unionFilters puts higher priority filters at the end", () => {
        expect(
            unionFilters(
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: "permenant",
                    },
                ],
                [
                    {
                        field: "bar",
                        operator: "in",
                        value: "crud",
                    },
                ],
                [
                    {
                        field: "baz",
                        operator: "in",
                        value: "prev",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "baz",
                "operator": "in",
                "value": "prev",
              },
              Object {
                "field": "bar",
                "operator": "in",
                "value": "crud",
              },
              Object {
                "field": "foo",
                "operator": "in",
                "value": "permenant",
              },
            ]
        `);
    });

    it("unionFilters should override same filters", () => {
        expect(
            unionFilters(
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: "permenant",
                    },
                ],
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: "crud",
                    },
                    {
                        field: "bar",
                        operator: "in",
                        value: "crud",
                    },
                ],
                [
                    {
                        field: "bar",
                        operator: "in",
                        value: "prev",
                    },
                    {
                        field: "baz",
                        operator: "in",
                        value: "prev",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "baz",
                "operator": "in",
                "value": "prev",
              },
              Object {
                "field": "bar",
                "operator": "in",
                "value": "crud",
              },
              Object {
                "field": "foo",
                "operator": "in",
                "value": "permenant",
              },
            ]
        `);
    });

    it("unionFilters should override even when filter value is null but should not keep it in the end result", () => {
        expect(
            unionFilters(
                [],
                [
                    {
                        field: "foo",
                        operator: "in",
                        value: null,
                    },
                    {
                        field: "bar",
                        operator: "in",
                        value: undefined,
                    },
                ],
                [
                    {
                        field: "bar",
                        operator: "in",
                        value: "prev",
                    },
                    {
                        field: "baz",
                        operator: "in",
                        value: "prev",
                    },
                ],
            ),
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "field": "baz",
                "operator": "in",
                "value": "prev",
              },
            ]
        `);
    });

    it("compareFilters filters are the same if their field and operator are the same", () => {
        expect(
            compareFilters(
                {
                    field: "foo",
                    operator: "in",
                    value: "left",
                },
                {
                    field: "foo",
                    operator: "in",
                    value: "right",
                },
            ),
        ).toBe(true);

        expect(
            compareFilters(
                {
                    field: "foo",
                    operator: "in",
                    value: "test",
                },
                {
                    field: "foo",
                    operator: "contains",
                    value: "test",
                },
            ),
        ).toBe(false);
    });
});
