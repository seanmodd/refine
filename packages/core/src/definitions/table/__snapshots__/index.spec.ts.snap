// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions/table stringify table params correctly 1`] = `"current=1&pageSize=10&categoryId__in[]=1&categoryId__in[]=2&sort[]=id&sort[]=title&order[]=desc&order[]=desc"`;

exports[`definitions/table stringify table single sort params correctly 1`] = `"current=1&pageSize=10&categoryId__in[]=1&categoryId__in[]=2&sort[]=id&order[]=desc"`;
